source/extensions/filters/common/rbac/matcher_extension.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/protobuf/message_validator.h"
+
+#include "source/extensions/filters/common/rbac/matchers.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RBAC {
+
+// Matcher extension factory for RBAC filter. Matchers could be extended to support Ip address,
+// header value etc.
+class MatcherExtensionFactory : public Envoy::Config::TypedFactory {
+public:
+  /**
+   * Function to create Matchers from the specified config.
+   * @param config supplies the matcher configuration
+   * @return a new MatcherExtension
+   */
+  virtual MatcherConstSharedPtr create(const Protobuf::Message& config,
+                                       ProtobufMessage::ValidationVisitor& validation_visitor) PURE;
+
+  // @brief the category of the matcher extension type for factory registration.
+  std::string category() const override { return "envoy.rbac.matchers"; }
+};
+
+// Base RBAC matcher extension factory. This facilitates easy creation of matcher extension
+// factories. The factory is templated by:
+//  M: Matcher extension implementation
+//  P: Protobuf definition of the matcher.
+template <typename M, typename P>
+class BaseMatcherExtensionFactory : public Filters::Common::RBAC::MatcherExtensionFactory {
+public:
+  Filters::Common::RBAC::MatcherConstSharedPtr
+  create(const Protobuf::Message& config,
+         ProtobufMessage::ValidationVisitor& validation_visitor) override {
+    const auto& matcher_typed_config =
+        MessageUtil::downcastAndValidate<const envoy::config::core::v3::TypedExtensionConfig&>(
+            config, validation_visitor);
```

Comment:
This `downcastAndValidate` doesn't validate the typed config, it should be done after anyConvert with `downcastAndValidate<const P&>`

Reply:
Thanks @lizan .Not sure I follow. Is there an example?

Reply:
Lizan I assume MessageUtil::anyConvertAndValidate would do both?

Reply:
Yes, but usually `create` takes a parameter after anyConvert. See network filter factory base as example:
https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/network/common/factory_base.h#L22-L27